server/connector/manager.go

```diff
@@ -106,33 +104,62 @@ func (m *manager) Filter(filters map[string]string) (subscribers []Subscriber) {
 
 func (m *manager) Add(s Subscriber) error {
 	logger.WithField("subscriber", s).WithField("lock", m.RWMutex).Debug("Add subscriber before locking")
-	m.Lock()
-	logger.WithField("subscriber", s).WithField("lock", m.RWMutex).Debug("Add subscriber lock acquired")
-	if _, found := m.subscribers[s.Key()]; found {
-		return ErrSubscriberExists
+	err := m.checkSubscribersExist(s)
+	if err != nil {
+		return err
 	}
-	m.Unlock()
 
 	if err := m.updateStore(s); err != nil {
 		return err
 	}
 
-	m.Lock()
-	m.subscribers[s.Key()] = s
-	m.Unlock()
+	m.putSubscriberInMap(s)
 	logger.WithField("subscriber", s).Debug("Add subscriber after updating store")
 	return nil
 }
 
 func (m *manager) Update(s Subscriber) error {
-	m.Lock()
-	defer m.Unlock()
+	err := m.checkSubscriberDoesNotExists(s)
+	if err != nil {
+		return err
+	}
+
+	m.putSubscriberInMap(s)
+	return m.updateStore(s)
+}
+
+func (m *manager) checkSubscriberDoesNotExists(s Subscriber) error {
+	m.RLock()
+	defer m.RUnlock()
+
 	if _, found := m.subscribers[s.Key()]; !found {
 		return ErrSubscriberDoesNotExist
 	}
 
+	return nil
+}
+
+func (m *manager) checkSubscribersExist(s Subscriber) error {
```

Comment:
these 2 related functions can be unified into a single func returning a boolean not an error
and their calling functions (upper level) can return the different errors Err..Exists, Err...NotExist